src/os_pcap_nodelet.cpp

```diff
@@ -0,0 +1,252 @@
+                read_packets(*pcap, pf);
+            }
+            NODELET_DEBUG("packet_read_thread done.");
+        });
+    }
+
+    void stop_packet_read_thread() {
+        NODELET_DEBUG("packet_read_thread stopping.");
+        if (packet_read_thread->joinable()) {
+            packet_read_active = false;
+            packet_read_thread->join();
+        }
+    }
+
+    void start_packet_processing_threads() {
+        imu_packets_processing_thread_active = true;
+        imu_packets_processing_thread = std::make_unique<std::thread>([this]() {
+            while (imu_packets_processing_thread_active) {
+                imu_packets->read([this](const uint8_t* buffer) {
+                    on_imu_packet_msg(buffer);
+                });
+            }
+            NODELET_DEBUG("imu_packets_processing_thread done.");
+        });
+
+        lidar_packets_processing_thread_active = true;
+        lidar_packets_processing_thread =
+            std::make_unique<std::thread>([this]() {
+                while (lidar_packets_processing_thread_active) {
+                    lidar_packets->read([this](const uint8_t* buffer) {
+                        on_lidar_packet_msg(buffer);
+                    });
+                }
+
+                NODELET_DEBUG("lidar_packets_processing_thread done.");
+            });
+    }
+
+    void stop_packet_processing_threads() {
+        NODELET_DEBUG("stopping packet processing threads.");
+
+        if (imu_packets_processing_thread->joinable()) {
+            imu_packets_processing_thread_active = false;
+            imu_packets_processing_thread->join();
+        }
+
+        if (lidar_packets_processing_thread->joinable()) {
+            lidar_packets_processing_thread_active = false;
+            lidar_packets_processing_thread->join();
+        }
+    }
+
+    void on_lidar_packet_msg(const uint8_t* raw_lidar_packet) {
+        // copying the data from queue buffer into the message buffer
+        // this can be avoided by constructing an abstraction where
+        // OusterSensor has its own RingBuffer of PacketMsg but for
+        // now we are focusing on optimizing the code for OusterDriver
+        std::memcpy(lidar_packet.buf.data(), raw_lidar_packet,
+                    lidar_packet.buf.size());
+        lidar_packet_pub.publish(lidar_packet);
+    }
+
+    void on_imu_packet_msg(const uint8_t* raw_imu_packet) {
+        // copying the data from queue buffer into the message buffer
+        // this can be avoided by constructing an abstraction where
+        // OusterSensor has its own RingBuffer of PacketMsg but for
+        // now we are focusing on optimizing the code for OusterDriver
+        std::memcpy(imu_packet.buf.data(), raw_imu_packet,
+                    imu_packet.buf.size());
+        imu_packet_pub.publish(imu_packet);
+    }
+
+    void read_packets(PcapReader& pcap, const sensor::packet_format& pf) {
+        size_t payload_size = pcap.next_packet();
```

Comment:
What happens if there are multiple sensors in this pcap?

Reply:
not a supported case.